### ISSUE TITLE
Revisit of this?

### ISSUE THREAD

**Issue by darksidelemm** [miweber67/spyserver_client#4]
Hi!

It seems that this repository is pretty much the only instance of a command-line spyserver client in existence. (Kind of surprising that that's the case, but it is!)

While I'm not really sure I want to use Spyservers as a primary backend for auto_rx (though it would be nice to have the option), there are certainly other applications where it would be useful to have a command-line SpyServer client.

For example, I've recently set up a few SpyServers at a very low-noise HF site, and it would be great to be able to connect to the the SpyServers locally on the site, and feed audio into something like direwolf for HF APRS reception, amongst other things.
Unfortunately, I haven't had much success in getting ss_client working with this spyserver. I'm able to connect, but I had to override the sample rate selection, and i'm not getting any samples through.

Any thoughts? I can send you the details for the spyserver if you like...

**Comment by darksidelemm** [comment]
Investigating further (lots of useful debug lines to uncomment and show the internal state - thanks!) I've found that the client didn't like my having bandwidth limits set on the server (I think it resulted in the number of decimation stages being too low, so the sample rate couldn't be set correctly).
Anyway, after removing the restrictions on the server, I was able to connect, but I'm noticing that the FIFO is filling up very slowly.

Example debug output is below. There's approximately one second between each of those "waiting with X samples" lines.

darkside$ ./ss_iq -r emailmeforserver.org -q 5040 -f 7100000 -s 12000 -b 16 -
bins for res: 120   fft bins: 128   resolution: 93.75Hz
SS_client_if(emailmeforserver.org, 5040)
SS_client_if: Trying to connect
SS_client_if: Connected

**********
Device Info:
   Type:                 2
   Serial:               875704369
   MaximumSampleRate:    768000
   MaximumBandwidth:     660000
   DecimationStageCount: 8
   GainStageCount:       0
   MaximumGainIndex:     0
   MinimumFrequency:     0
   MaximumFrequency:     1700000000
   Resolution:           16
   MinimumIQDecimation:  0
   ForcedIQFormat:       0

**********
Client sync:
   Control:     No
   gain:        5
   dev_ctr:     7100000
   ch_ctr:      7100000
   iq_ctr:      7100000
   fft_ctr:     7100000
   min_iq_ctr:  6770000
   max_iq_ctr:  7430000
   min_fft_ctr: 7100000
   max_fft_ctr: 7100000

**********
Client sync:
   Control:     No
   gain:        5
   dev_ctr:     7100000
   ch_ctr:      7100000
   iq_ctr:      7100000
   fft_ctr:     7100000
   min_iq_ctr:  7100000
   max_iq_ctr:  7100000
   min_fft_ctr: 7100000
   max_fft_ctr: 7100000
SS_client_if: Supported Sample Rates:
SS_client_if:   3000		8
SS_client_if:   6000		7
SS_client_if:   12000		6
SS_client_if:   24000		5
SS_client_if:   48000		4
SS_client_if:   96000		3
SS_client_if:   192000		2
SS_client_if:   384000		1
SS_client_if:   768000		0
SS_client_if: Ready
Exact decimation match
Desired decimation stage: 6 (768000 / 64 = 12000) resample ratio: 1
ss_client: setting center_freq to 7.1e+06
SS_client_if: Sample Rate: 12000
SS_client_if: Setting IQ sample rate to 12000 stage 6

**********
Client sync:
   Control:     No
   gain:        5
   dev_ctr:     7100000
   ch_ctr:      7100000
   iq_ctr:      7100000
   fft_ctr:     7100000
   min_iq_ctr:  6775156
   max_iq_ctr:  7424844
   min_fft_ctr: 7100000
   max_fft_ctr: 7100000
SS_client_if: The server does not allow you to change the gains.
SS_client_if: Starting Streaming
ss_client_if::get_iq_data: sizeof(T)=2
batch size: 32768   iq_samp_rate: 12000
max_wait for iq data check is 910ms
   Waiting with 0 samples available, want 32768
IN: tail	0	head	0	free	10485760
   Waiting with 120 samples available, want 32768
IN: tail	0	head	480	free	10485280
   Waiting with 240 samples available, want 32768
IN: tail	0	head	960	free	10484800
   Waiting with 360 samples available, want 32768
IN: tail	0	head	1440	free	10484320
IN: tail	0	head	1920	free	10483840
   Waiting with 600 samples available, want 32768
IN: tail	0	head	2400	free	10483360
IN: tail	0	head	2880	free	10482880

**Comment by miweber67** [comment]
@darksidelemm Sorry for the delay in getting back to you! Things have been a little busy. I will look in to this and let you know what I find.

**Comment by darksidelemm** [comment]
Hi! If you shoot me an email at vk5qi@rfhead.net I can send you some details for the server which this issue occurred on.

Cheers,
Mark

**Comment by miweber67** [comment]
> It seems that this repository is pretty much the only instance of a command-line spyserver client in existence. (Kind of surprising that that's the case, but it is!)

I know, right?!? If I could have found an alternative, I wouldn't have written this one. :-)

Email sent.

**Comment by miweber67** [comment]
OK, update: I fixed two bugs relating to decimation stage selection; one in the client where I had a leftover hard-coded request for stage 5 which was completely unnecessary, and one in the interface where all indices were assumed to be present in the list, which they are not when BW limits are engaged. So that should work now.

As far as the FIFO fill rate, I did not observe that problem this morning. Data flowed in from all three of your servers at about the expected rate; each packet contained 120 samples as in your logs. I captured one stream in wireshark and verified that the incoming data rate was a little more than 48KBps, which is what you should get with 12ksps at 16 bits per complex sample (12000 * 2 * 2).

So at this point, I cannot identify anything in the ss_client that would cause the behavior you're observing, possibly due to differences in my runtime environment from yours. I am happy to try additional tests if you can identify any that may help, but at the moment I don't have any ideas.

**Comment by darksidelemm** [comment]
Exactly what command were you using? (Leaving out the hostname..)

**Comment by darksidelemm** [comment]
I should also note that I'm getting errors building this on a Raspberry Pi (GCC 8.3)
```
pi@satnogs2:~/spyserver_client $ make
g++ -I. -O3 -Wall   -c -o ss_client.o ss_client.cc
g++ -I. -O3 -Wall   -c -o tcp_client.o tcp_client.cc
g++ -I. -O3 -Wall   -c -o ss_client_if.o ss_client_if.cc
In file included from /usr/include/c++/8/bits/stl_algo.h:61,
                 from /usr/include/c++/8/algorithm:62,
                 from ss_client_if.cc:10:
/usr/include/c++/8/bits/stl_heap.h: In function ‘void std::__adjust_heap(_RandomAccessIterator, _Distance, _Distance, _Tp, _Compare) [with _RandomAccessIterator = __gnu_cxx::__normal_iterator<std::pair<double, unsigned int>*, std::vector<std::pair<double, unsigned int> > >; _Distance = int; _Tp = std::pair<double, unsigned int>; _Compare = __gnu_cxx::__ops::_Iter_less_iter]’:
/usr/include/c++/8/bits/stl_heap.h:214:5: note: parameter passing for argument of type ‘__gnu_cxx::__normal_iterator<std::pair<double, unsigned int>*, std::vector<std::pair<double, unsigned int> > >’ changed in GCC 7.1
     __adjust_heap(_RandomAccessIterator __first, _Distance __holeIndex,
     ^~~~~~~~~~~~~
In file included from /usr/include/c++/8/algorithm:62,
                 from ss_client_if.cc:10:
/usr/include/c++/8/bits/stl_algo.h: In function ‘void std::__introsort_loop(_RandomAccessIterator, _RandomAccessIterator, _Size, _Compare) [with _RandomAccessIterator = __gnu_cxx::__normal_iterator<std::pair<double, unsigned int>*, std::vector<std::pair<double, unsigned int> > >; _Size = int; _Compare = __gnu_cxx::__ops::_Iter_less_iter]’:
/usr/include/c++/8/bits/stl_algo.h:1940:5: note: parameter passing for argument of type ‘__gnu_cxx::__normal_iterator<std::pair<double, unsigned int>*, std::vector<std::pair<double, unsigned int> > >’ changed in GCC 7.1
     __introsort_loop(_RandomAccessIterator __first,
     ^~~~~~~~~~~~~~~~
/usr/include/c++/8/bits/stl_algo.h:1940:5: note: parameter passing for argument of type ‘__gnu_cxx::__normal_iterator<std::pair<double, unsigned int>*, std::vector<std::pair<double, unsigned int> > >’ changed in GCC 7.1
/usr/include/c++/8/bits/stl_algo.h:1954:25: note: parameter passing for argument of type ‘__gnu_cxx::__normal_iterator<std::pair<double, unsigned int>*, std::vector<std::pair<double, unsigned int> > >’ changed in GCC 7.1
    std::__introsort_loop(__cut, __last, __depth_limit, __comp);
    ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/8/bits/stl_algo.h: In function ‘void std::__insertion_sort(_RandomAccessIterator, _RandomAccessIterator, _Compare) [with _RandomAccessIterator = __gnu_cxx::__normal_iterator<std::pair<double, unsigned int>*, std::vector<std::pair<double, unsigned int> > >; _Compare = __gnu_cxx::__ops::_Iter_less_iter]’:
/usr/include/c++/8/bits/stl_algo.h:1840:5: note: parameter passing for argument of type ‘__gnu_cxx::__normal_iterator<std::pair<double, unsigned int>*, std::vector<std::pair<double, unsigned int> > >’ changed in GCC 7.1
     __insertion_sort(_RandomAccessIterator __first,
     ^~~~~~~~~~~~~~~~
/usr/include/c++/8/bits/stl_algo.h:1840:5: note: parameter passing for argument of type ‘__gnu_cxx::__normal_iterator<std::pair<double, unsigned int>*, std::vector<std::pair<double, unsigned int> > >’ changed in GCC 7.1
In file included from /usr/include/c++/8/vector:69,
                 from ss_client_if.h:16,
                 from ss_client_if.cc:17:
/usr/include/c++/8/bits/vector.tcc: In member function ‘void std::vector<_Tp, _Alloc>::_M_realloc_insert(std::vector<_Tp, _Alloc>::iterator, _Args&& ...) [with _Args = {std::pair<double, unsigned int>}; _Tp = std::pair<double, unsigned int>; _Alloc = std::allocator<std::pair<double, unsigned int> >]’:
/usr/include/c++/8/bits/vector.tcc:413:7: note: parameter passing for argument of type ‘std::vector<std::pair<double, unsigned int> >::iterator’ {aka ‘__gnu_cxx::__normal_iterator<std::pair<double, unsigned int>*, std::vector<std::pair<double, unsigned int> > >’} changed in GCC 7.1
       vector<_Tp, _Alloc>::
       ^~~~~~~~~~~~~~~~~~~
/usr/include/c++/8/bits/vector.tcc: In member function ‘void ss_client_if::on_connect()’:
/usr/include/c++/8/bits/vector.tcc:109:4: note: parameter passing for argument of type ‘__gnu_cxx::__normal_iterator<std::pair<double, unsigned int>*, std::vector<std::pair<double, unsigned int> > >’ changed in GCC 7.1
    _M_realloc_insert(end(), std::forward<_Args>(__args)...);
    ^~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/8/algorithm:62,
                 from ss_client_if.cc:10:
/usr/include/c++/8/bits/stl_algo.h:1968:25: note: parameter passing for argument of type ‘__gnu_cxx::__normal_iterator<std::pair<double, unsigned int>*, std::vector<std::pair<double, unsigned int> > >’ changed in GCC 7.1
    std::__introsort_loop(__first, __last,
    ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~
     std::__lg(__last - __first) * 2,
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     __comp);
     ~~~~~~~
/usr/include/c++/8/bits/stl_algo.h:1885:25: note: parameter passing for argument of type ‘__gnu_cxx::__normal_iterator<std::pair<double, unsigned int>*, std::vector<std::pair<double, unsigned int> > >’ changed in GCC 7.1
    std::__insertion_sort(__first, __first + int(_S_threshold), __comp);
    ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/8/bits/stl_algo.h:1890:23: note: parameter passing for argument of type ‘__gnu_cxx::__normal_iterator<std::pair<double, unsigned int>*, std::vector<std::pair<double, unsigned int> > >’ changed in GCC 7.1
  std::__insertion_sort(__first, __last, __comp);
  ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
g++ -o ss_client ss_client.o tcp_client.o ss_client_if.o -I. -O3 -Wall -lpthread -lsamplerate
/usr/bin/ld: ss_client_if.o: in function `ss_client_if::disconnect()':
ss_client_if.cc:(.text+0x67c): undefined reference to `__atomic_store_8'
/usr/bin/ld: ss_client_if.o: in function `ss_client_if::cleanup()':
ss_client_if.cc:(.text+0x73c): undefined reference to `__atomic_store_8'
/usr/bin/ld: ss_client_if.o: in function `ss_client_if::parse_message(char*, unsigned int)':
ss_client_if.cc:(.text+0x14a4): undefined reference to `__atomic_fetch_add_8'
/usr/bin/ld: ss_client_if.o: in function `ss_client_if::thread_loop()':
ss_client_if.cc:(.text+0x1bec): undefined reference to `__atomic_store_8'
/usr/bin/ld: ss_client_if.o: in function `ss_client_if::~ss_client_if()':
ss_client_if.cc:(.text+0x1e28): undefined reference to `__atomic_store_8'
/usr/bin/ld: ss_client_if.o: in function `ss_client_if::start()':
ss_client_if.cc:(.text+0x1fec): undefined reference to `__atomic_store_8'
/usr/bin/ld: ss_client_if.o: in function `ss_client_if::stop()':
ss_client_if.cc:(.text+0x20dc): undefined reference to `__atomic_store_8'
/usr/bin/ld: ss_client_if.o: in function `ss_client_if::connect()':
ss_client_if.cc:(.text+0x3d34): undefined reference to `__atomic_store_8'
collect2: error: ld returned 1 exit status
make: *** [makefile:11: ss_client] Error 1
```

**Comment by miweber67** [comment]
I emailed you details of my tests.

With respect to the compilation issue, please try adding `-latomic` to the link line:

`g++ -o ss_client ss_client.o tcp_client.o ss_client_if.o -I. -O3 -Wall -lpthread -lsamplerate -latomic`

I am using g++ 8.3.0 on x86-64 to build; will have to try the pi later but cannot do it now.

**Comment by darksidelemm** [comment]
OK, I've had some intermittent success running on a RPi. The stream takes a little while to get going, and writes out in rather large chunks. 

I've been able to get it running (kind of) under OSX (had to use --std=c++11 to get clang to work nicely), though it sometimes breaks when samples are dropped. I don't see the similar dropped samples on the RPi, but that may be intermittent network issues. 

I've also noted that I can't tune away from the centre frequency that the SDR is locked to. E.g. the SDR is locked to a centre freq of 7.1 MHz, and while SDR# lets me tune around the entire receive passband (e.g. 7.1 MHz +/- 768000/2), ss_client will lock me to 7.1 MHz.

**Comment by darksidelemm** [comment]
On the point of the frequency range, I see that the spyserver is not being very helpful here. I see multiple Client sync messages showing up during the connection process, some with:
   min_iq_ctr:  6770000
   max_iq_ctr:  7430000
(which  is correct)
... and then others with:
   min_iq_ctr:  7100000
   max_iq_ctr:  7100000
(which is wrong).

**Comment by miweber67** [comment]
> I've also noted that I can't tune away from the centre frequency that the SDR is locked to. E.g. the SDR is locked to a centre freq of 7.1 MHz, and while SDR# lets me tune around the entire receive passband (e.g. 7.1 MHz +/- 768000/2), ss_client will lock me to 7.1 MHz.

OK, I think I have worked around the spyserver init process so that you should be able to select an IQ center that is within the locked passband of a served SDR. Ironically, I had to replace the call I removed earlier with the fixed decimation of '5', only this time I use the actual desired decimation factor. Otherwise spyserver will not report the correct bounds. As part of this workaround, I have put a 2-second sleep in to allow time for the spyserver to return the updated client sync block, without which we cannot set the desired freq. If this works for you I will put a more intelligent wait on the to-do list.

> The stream takes a little while to get going, and writes out in rather large chunks.

I added a -a option so you can change the chunk size. If you set it to 1200 you should get ten per second at your 12000 BW rate.

`./ss_iq -a 1200 -r server.example.com -q 5555 -f 7101000 -s 12000 -b 16 5040_7101.bin`

I also added -latomic to the makefile, since it didn't hurt me here and it seems it's needed on the pi?

Please let me know if these changes  work for you.

**Comment by darksidelemm** [comment]
Yep, this is working!! Thanks!

The only major issue now is the latency through the system, which I measure at about 2 seconds (for reference I get about 400ms latency from SDR#). I'm wondering if the 2 second sleep is contributing to this?

**Comment by miweber67** [comment]
OK, progress!
I doubt the 2 second sleep is the origin of the latency, but you can test this by adjusting the value in ss_client.cc line 465:

`std::this_thread::sleep_for(std::chrono::milliseconds(2000));`

The only other thing I can think of offhand is that stdout may be buffered; you can force each write to be flushed by adding an `out->flush()` after the write in ss_client.cc line 563 (for 16-bit samples):
```
            out->write(buf, data.output_frames_gen*2*2);
            out->flush();
```
I tested this by piping stdout to xxd and I did not observe any difference, but I do not have the same reference point for measuring latency that you do. How are you measuring latency? What is the full pipeline of the system?

**Comment by darksidelemm** [comment]
The SDR I'm receiving on is within ground-wave distance of my home station. 
I can transmit a CW dit from my home station, and time how long it takes to be heard through the SDR chain.

I'll do some of the tests you mention and see if it changes anything.

**Comment by darksidelemm** [comment]
OK, the issue is probably the processing in my USB demodulation chain. Will need to look at lowering the filter lengths.

**Comment by darksidelemm** [comment]
So trying out just listening to the IQ audio directly (pseudo-DSB receiver), the latency is still around the 1 second mark, but variable. I'm wondering if there's some data getting loaded up in the TCP receiver buffer or something that needs to be cleared or something...

**Comment by miweber67** [comment]
>I'm wondering if there's some data getting loaded up in the TCP receiver buffer or something that needs to be cleared or something... 

That is a very good thought. Try adjusting this value in ss_client_if.h to something smaller than the current value, which is:
```
   static constexpr unsigned int BufferSize = 64 * 1024;
```

**Comment by darksidelemm** [comment]
I've tried lower that value a fair way - it looks like i might be hitting some latency limit to do with audio processing - around 1 second. 

Honestly, 1 second latency is *probably* OK for a lot of the applications we are interested in (HF APRS, etc). Its probably pushing it for time-critical modes like FT8, but we have other receiver solutions in mind for those. 

Thanks so much for your work on this - With what you've got going so far, I think i'm not far off being able to run a multi-band HF APRS iGate using our remote receive site.

**Comment by miweber67** [comment]
Cool! Good luck and please let me know how it all works out. If you need anything else feel free to open up another issue or send me an email.
